docs/server datastorage.md

```diff
@@ -0,0 +1,173 @@
+        {"operation": "max", "value": 0},
+    ]
+}
+```
+Response:
+```json
+{
+    "cmd": "Set",
+    "key": "EnergyLink0",
+    "tag": "7cc04194-b491-4cba-a89e-ef754502f3ff",
+    "default": 0,
+    "slot": 5,
+    "want_reply": true,
+    "operations": [
+        {"operation": "add", "value": -50},
+        {"operation": "max", "value": 0},
+    ],
+    "original_value": 20,
+    "value": 0
+
+}
+```
+So we only where able to subtract `original_value` - `value` = 20 Energy.
+
+### Gifting
+Gifting allows players to gift items they don't need to other games, like if you get a health potion but you are at no risk of dieing someone else might have more use for it. Gifting works by slot specific giftboxes, any client can add a gift to the giftbox, and only the client of the slot the giftbox belongs to is supposed to remove gifts from it. Now it might seem easy with an array of gift objects however while appending to an array is thread safe, removing entries from an array cannot be done thread safe. Therefor giftboxes are implemented as dictionaries where each gift has its own unique key. new gifts can safely be added using the `update` operation and removal of specific gifts can be done using their unique keys with the `pop` operation. (for more imformation visit the [gifting api documentation](https://github.com/agilbert1412/Archipelago.Gifting.Net/blob/main/Documentation/Gifting%20API.md)).
+
+Adding gifts:
+```json
+{
+    "cmd": "Set",
+    "key": "GiftBox;0;1",
+    "default": {},
+    "operations": [
+        {"operation": "update", "value": {
+            "1a4a169f-de12-45cf-a7d9-f77c1c2ebc31": {
+                "gift": "json"
+            }
+        }}
+    ]
+}
+```
+Removing gifts, note we are ignoring the errors here as pop-ing an non existing key will raise an error otherwise.
+```json
+{
+    "cmd": "Set",
+    "key": "GiftBox;0;1",
+    "default": {},
+    "on_error": "ignore",
+    "operations": [
+        {"operation": "pop", "value": "1a4a169f-de12-45cf-a7d9-f77c1c2ebc31" },
+    ]
+}
+```
+
+### Queue (FIFO) \ Stack (FILO) 
+Queues can be implemented using lists and this will still be safe aslong as all producers only append to the end of the list and consumers only remove the first or last element in the list. To append to a list we would use the `add` operation with an array value, this will append all elements in the provided array to the list. In order to `peek` any elemment in the list, you would unfortunately have to retrieve the full list by its key. To remove the first or last element of a list we can use the `pop` opperation.
+
+Adding elements to the list is easy, just a `Set`-`add` operation with an array as value:
+```json
+{
+    "cmd": "Set",
+    "key": "SomeQueueName",
+    "default": [],
+    "operations": [
+        {"operation": "add", "value": ["whatever value"] },
+    ]
+}
+```
+To pop an element in the queue we would use a `Set`-`pop` operation, with a value of `0` for queues and `1` for stacks

```

Comment:
Pretty sure this part about stacks is wrong, it will pop the item at index 1, not the last item.

The correct implementation of a stack in datastorage is probably to insert at index 0 (if possible) and then pop from index 0